src/lib/profiles/data-management/Current/SubscriptionEngine.cpp

```diff
@@ -1515,7 +1520,675 @@ void SubscriptionEngine::OnCustomCommand(nl::Weave::ExchangeContext * aEC, const
+            }
+            elementTracker++;
+        }
+        lVLBuilder.EndOfVersionList();
+        SuccessOrExit(lVLBuilder.GetError());
+    }
+
+    elementTracker = static_cast<struct StatusDataHandleElement *>(context->mpFirstStatusDataHandleElement);
+    {
+        StatusList::Builder & lSLBuilder = updateResponseBuilder.CreateStatusListBuilder();
+        for (uint32_t j = 0; j < context->mNumDataElements; j++)
+        {
+            lSLBuilder.AddStatus(elementTracker->mProfileId, elementTracker->mStatusCode);
+            elementTracker++;
+        }
+        lSLBuilder.EndOfStatusList();
+        SuccessOrExit(lSLBuilder.GetError());
+    }
+
+    updateResponseBuilder.EndOfResponse();
+    SuccessOrExit(updateResponseBuilder.GetError());
+
+    aWriter.Finalize();
+
+    WeaveLogDetail(DataManagement, "ConstructStatusListVersionList success with number of elements %d", context->mNumDataElements);
+
+exit:
+    if (err != WEAVE_NO_ERROR)
+    {
+        aWriter = checkpoint;
+    }
+}
+
+void SubscriptionEngine::BuildStatusDataHandleElement(PacketBuffer * pBuf, TraitDataHandle aTraitDataHandle, WEAVE_ERROR & err,
+                                                      uint32_t aCurrentIndex)
+{
+    uint32_t profileId  = 0;
+    uint16_t statusCode = 0;
+
+    if (WEAVE_ERROR_ACCESS_DENIED == err)
+    {
+        profileId  = Weave::Profiles::kWeaveProfile_Common;
+        statusCode = Weave::Profiles::Common::kStatus_AccessDenied;
+        err        = WEAVE_NO_ERROR;
+    }
+    else if (WEAVE_ERROR_INVALID_PROFILE_ID == err)
+    {
+        profileId  = Weave::Profiles::kWeaveProfile_WDM;
+        statusCode = kStatus_InvalidPath;
+        err        = WEAVE_NO_ERROR;
+    }
+    else if (WEAVE_ERROR_WDM_VERSION_MISMATCH == err)
+    {
+        profileId  = Weave::Profiles::kWeaveProfile_WDM;
+        statusCode = kStatus_VersionMismatch;
+        err        = WEAVE_NO_ERROR;
+    }
+    else if (WEAVE_NO_ERROR == err)
+    {
+        profileId  = Weave::Profiles::kWeaveProfile_Common;
+        statusCode = Weave::Profiles::Common::kStatus_Success;
+    }
+    else
+    {
+        profileId  = Weave::Profiles::kWeaveProfile_Common;
+        statusCode = Weave::Profiles::Common::kStatus_InternalError;
+    }
+
+    StatusDataHandleElement * statusDataHandleList =
+        reinterpret_cast<StatusDataHandleElement *> WEAVE_SYSTEM_ALIGN_SIZE((size_t)(pBuf->Start()), 4);
+    statusDataHandleList[aCurrentIndex].mProfileId       = profileId;
+    statusDataHandleList[aCurrentIndex].mStatusCode      = statusCode;
+    statusDataHandleList[aCurrentIndex].mTraitDataHandle = aTraitDataHandle;
+}
+
+/**
+ * Check if this triat path is the starting one
+ */
+bool SubscriptionEngine::IsStartingPath(StatusDataHandleElement * apStatusDataHandleList, TraitDataHandle aTraitDataHandle,
+                                        uint32_t aCurrentIndex)
+{
+    bool isStarting = true;
+    // Todo: Optimize to reduce loop lookup
```

Comment:
a convention of all caps TODO is slightly more easily searched, but a non-blocker here

Reply:
Done, thanks